server-state-machine.go

```diff
@@ -129,9 +139,29 @@ func (state ServerStateStart) Next(hm *HandshakeMessage) (HandshakeState, []Hand
 		return nil, nil, AlertProtocolVersion
 	}
 
-	if state.Caps.RequireCookie && state.cookieSent && !state.Caps.CookieHandler.Validate(state.conn, clientCookie.Cookie) {
-		logf(logTypeHandshake, "[ServerStateStart] Cookie mismatch")
-		return nil, nil, AlertAccessDenied
+	// The client sent a cookie. So this is probably the second ClientHello (sent as a response to a HRR)
+	var firstClientHello *HandshakeMessage
+	if clientSentCookie {
+		plainCookie, err := state.Caps.CookieSource.DecodeToken(clientCookie.Cookie)
+		if err != nil {
+			logf(logTypeHandshake, fmt.Sprintf("[ServerStateStart] Error decoding token [%v]", err))
+			return nil, nil, AlertAccessDenied
```

Comment:
I don't think you want AccessDenied here, but rather DecryptError or IllegalParameter

Reply:
I would put a TODO here because if you were to change your key, this wouldn't work.